lib/Target/TfheRustHL/TfheRustHLEmitter.cpp

```diff
@@ -0,0 +1,653 @@
+  std::function<std::string(Value)> valueOrClonedValue = [&](Value value) {
+    if (isa<BlockArgument>(value)) {
+      // Function arguments used as outputs must be cloned.
+      return variableNames->getNameForValue(value) + ".clone()";
+    } else if (MemRefType memRefType = dyn_cast<MemRefType>(value.getType())) {
+      auto shape = memRefType.getShape();
+      // Internally allocated memrefs that are treated as hashmaps must be
+      // converted to arrays.
+      unsigned int i = 0;
+      std::string res =
+          variableNames->getNameForValue(value) + std::string(".get(&(") +
+          std::accumulate(std::next(shape.begin()), shape.end(),
+                          std::string("i0"),
+                          [&](std::string a, int64_t value) {
+                            return a + ", i" + std::to_string(++i);
+                          }) +
+          std::string(")).unwrap().clone()");
+      for ([[maybe_unused]] unsigned _ : shape) {
+        res = llvm::formatv("core::array::from_fn(|i{0}| {1})", i--, res);
+      }
+      return res;
+    }
+    return variableNames->getNameForValue(value);
+  };
+
+  if (op.getNumOperands() == 0) {
+    return success();
+  }
+
+  if (op.getNumOperands() == 1) {
+    os << valueOrClonedValue(op.getOperands()[0]) << "\n";
+    return success();
+  }
+
+  os << "(" << commaSeparatedValues(op.getOperands(), valueOrClonedValue)
+     << ")\n";
+  return success();
+}
+
+void TfheRustHLEmitter::emitAssignPrefix(Value result) {
+  os << "let " << variableNames->getNameForValue(result) << " = ";
+}
+
+LogicalResult TfheRustHLEmitter::printMethod(
+    ::mlir::Value result, ::mlir::ValueRange nonSksOperands,
+    std::string_view op, SmallVector<std::string> operandTypes) {
+  emitAssignPrefix(result);
+
+  auto *operandTypesIt = operandTypes.begin();
+  // os << variableNames->getNameForValue(sks) << "." << op << "(";
+  os << commaSeparatedValues(nonSksOperands, [&](Value value) {
+    auto *prefix = value.getType().hasTrait<PassByReference>() ? "&" : "";
+    // First check if a DefiningOp exists
+    // if not: comes from function definition
+    mlir::Operation *op = value.getDefiningOp();
+    if (op) {
+      auto referencePredicate =
+          isa<tensor::ExtractOp>(op) || isa<memref::LoadOp>(op);
+      prefix = referencePredicate ? "" : prefix;
+    } else {
+      prefix = "";
+    }
+
+    return prefix + variableNames->getNameForValue(value) +
+           (!operandTypes.empty() ? " as " + *operandTypesIt++ : "");
+  });
+
+  os << ");\n";
+
+  return success();
+
+  return failure();
+}
+
+LogicalResult TfheRustHLEmitter::printOperation(CreateTrivialOp op) {
+  emitAssignPrefix(op.getResult());
+  os << "FheUint" << DefaultTfheRustHLBitWidth << "::try_encrypt_trivial("
+     << variableNames->getNameForValue(op.getValue()) << ").unwrap();\n";
+  return success();
+}
+
+LogicalResult TfheRustHLEmitter::printOperation(affine::AffineForOp op) {
```

Comment:
For some of these I wonder if it would be worthwhile making a base class so that all the Tfhe-RS emitters can share these common implementations. Maybe file an issue for future work?